gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/ContentsUtil.java

```diff
@@ -16,65 +16,63 @@
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
-import java.util.Optional;
 import java.util.concurrent.Callable;
-import java.util.concurrent.Future;
 import java.util.stream.Collectors;
+import org.immutables.value.Value;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public final class ContentsUtil {
     private static final Logger log = LoggerFactory.getLogger(ContentsUtil.class);
 
-    public static Optional<String> fetchPageContents(URL pageUrl) {
-        ProgressIndicator indicator = ProgressManager.getInstance().getProgressIndicator();
-
-        if (indicator == null) {
-            return Optional.empty();
+    public static ContentResults fetchPageContents(String urlString) {
+        try {
+            URL url = new URL(urlString);
+            return ContentsUtil.fetchPageContents(url);
+        } catch (MalformedURLException e) {
+            log.error("Malformed URL", e);
+            return ContentResults.empty();
         }
+    }
 
+    public static ContentResults fetchPageContents(URL pageUrl) {
         try {
-            Future<String> future =
-                    ApplicationManager.getApplication().executeOnPooledThread(fetchContentTask(pageUrl, indicator));
-            String content =
-                    com.intellij.openapi.application.ex.ApplicationUtil.runWithCheckCanceled(future::get, indicator);
-            return Optional.ofNullable(content);
+            return fetchContentTask(pageUrl).call();
         } catch (InterruptedException | ProcessCanceledException e) {
-            log.debug("Fetch operation was cancelled", e);
+            log.warn("Fetch operation was cancelled", e);
+            return ContentResults.empty();
         } catch (Exception e) {
             log.warn("Failed to fetch contents", e);
+            return ContentResults.empty();
         }
-        return Optional.empty();
     }
 
-    private static Callable<String> fetchContentTask(URL pageUrl, ProgressIndicator indicator) {
+    private static Callable<ContentResults> fetchContentTask(URL pageUrl) {
         return () -> {
             HttpURLConnection connection = null;
             try {
                 connection = (HttpURLConnection) pageUrl.openConnection();
                 connection.setRequestMethod("GET");
 
-                if (indicator.isCanceled() || connection.getResponseCode() != 200) {
-                    throw new ProcessCanceledException();
+                int responseCode = connection.getResponseCode();
+                if (responseCode != HttpURLConnection.HTTP_OK) {
+                    return ContentResults.error(responseCode);
                 }
 
                 BufferedReader in =
                         new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
-
-                return in.lines().collect(Collectors.joining("\n"));
+                return ContentResults.of(in.lines().collect(Collectors.joining("\n")));
```

Comment:
Getting the lines then joining them back together is kinda too much effort. I would do `CharStreams.toString(inputStreamReader)`